pkg/pool-hooks/test/foundry/StableSurgeExample.t.sol

```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import "forge-std/Test.sol";
+
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IVaultAdmin } from "@balancer-labs/v3-interfaces/contracts/vault/IVaultAdmin.sol";
+import {
+    PoolRoleAccounts,
+    SwapKind,
+    PoolSwapParams
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { CastingHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/CastingHelpers.sol";
+import { ArrayHelpers } from "@balancer-labs/v3-solidity-utils/contracts/test/ArrayHelpers.sol";
+import { StablePoolFactory } from "@balancer-labs/v3-pool-stable/contracts/StablePoolFactory.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+import { StableMath } from "@balancer-labs/v3-solidity-utils/contracts/math/StableMath.sol";
+import { BaseVaultTest } from "@balancer-labs/v3-vault/test/foundry/utils/BaseVaultTest.sol";
+
+import { StableSurgeHookExample } from "../../contracts/StableSurgeHookExample.sol";
+
+contract StableSurgeHookExampleTest is BaseVaultTest {
+    using CastingHelpers for address[];
+    using ArrayHelpers for *;
+    using FixedPoint for uint256;
+
+    uint256 internal daiIdx;
+    uint256 internal usdcIdx;
+
+    StablePoolFactory internal stablePoolFactory;
+
+    uint64 SWAP_FEE_PERCENTAGE = 0.0004e18;
+    uint256 AMP_FACTOR = 200;
+    // A threshold of 0.1 for a 2 token pool would mean surging would occur if any token reaches 60% of the total of balances.
+    uint256 THRESHOLD = 0.1e18;
+    uint256 SURGE_COEFFICIENT = 50e18;
+
+    // The following reference values were calculated manually using settings above
+    uint256 initialAmount = 50e18;
+    uint256[] balancesScaled18 = [initialAmount, initialAmount].toMemoryArray();
+    uint256 expectedThresholdBoundary = 0.6e18;
+    uint256 amountInBelowThreshold = 1e18;
+    uint256 amountInAboveThreshold = 30e18;
+    // Return from GivenIn swap, input=amountInAboveThreshold, minus fee
+    uint256 amountOutAboveThreshold = 29.850106042102014697e18;
```

Comment:
These would be calculated in the code based on the constants above, but we can look at the tests once the main contract is done.